janus_core/src/retries.rs

```diff
@@ -0,0 +1,315 @@
+                    }
+                }
+
+                debug!("encountered non-retryable error");
+                Err(backoff::Error::permanent(Err(error)))
+            }
+        }
+    })
+    .await;
+
+    match result {
+        Ok(response) => Ok(response),
+        Err(result) => result,
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::test_util::install_test_trace_subscriber;
+    use mockito::mock;
+    use tokio::net::TcpListener;
+    use url::Url;
+
+    #[tokio::test]
+    async fn http_retry_client_error() {
+        install_test_trace_subscriber();
+
+        let mock_404 = mock("GET", "/")
+            .with_status(StatusCode::NOT_FOUND.as_u16().into())
+            .with_header("some-header", "some-value")
+            .with_body("some-body")
+            .expect(1)
+            .create();
+
+        let http_client = reqwest::Client::builder().build().unwrap();
+
+        // HTTP 404 should cause the client to give up after a single attempt, and the caller should
+        // get `Ok(reqwest::Response)`.
+        let response = retry_http_request(test_http_request_exponential_backoff(), || async {
+            http_client.get(mockito::server_url()).send().await
+        })
+        .await
+        .unwrap();
+
+        assert_eq!(response.status(), StatusCode::NOT_FOUND);
+        assert_eq!(
+            response.headers().get("some-header").unwrap(),
+            &"some-value"
+        );
+        assert_eq!(response.text().await.unwrap(), "some-body".to_string());
+
+        mock_404.assert();
+    }
+
+    #[tokio::test]
+    async fn http_retry_server_error() {
+        install_test_trace_subscriber();
+
+        let mock_500 = mock("GET", "/")
+            .with_status(StatusCode::INTERNAL_SERVER_ERROR.as_u16().into())
+            .with_header("some-header", "some-value")
+            .with_body("some-body")
+            .expect_at_least(2)
+            .create();
+
+        let http_client = reqwest::Client::builder().build().unwrap();
+
+        // We expect to eventually give up in the face of repeated HTTP 500, but the caller expects
+        // a `reqwest::Response` so they can examine the status code, headers and response body,
+        // which you can't get from a `reqwest::Error`.
+        let response = retry_http_request(test_http_request_exponential_backoff(), || async {
+            http_client.get(mockito::server_url()).send().await
+        })
+        .await
+        .unwrap();
+
+        assert_eq!(response.status(), StatusCode::INTERNAL_SERVER_ERROR);
+        assert_eq!(
+            response.headers().get("some-header").unwrap(),
+            &"some-value"
+        );
+        assert_eq!(response.text().await.unwrap(), "some-body".to_string());
+        mock_500.assert();
+    }
+
+    #[tokio::test]
+    async fn http_retry_server_err1or_unimplemented() {
+        install_test_trace_subscriber();
+
+        let mock_500 = mock("GET", "/")
```

Comment:
This is probably where that 1 was headed :)

```suggestion
        let mock_501 = mock("GET", "/")
```

Reply:
It's worse than that: I was just trying to force `cargo test`'s test name matcher to ignore those tests.